src/backend.rs

```diff
@@ -0,0 +1,465 @@
+                                            libevent_sys::evmap_io_active_(base.0, fd, libevent_sys::EV_READ as i16);
+                                        }
+
+                                        // If the ready flag is not cleared, then this loop will hang the runtime.
+                                        guard.clear_ready();
+                                    }
+                                    Err(error) => {
+                                        tracing::error!(?error);
+                                        break;
+                                    }
+                                }
+                            },
+                            result = async_fd.writable(), if is_write => {
+                                match result {
+                                    Ok(mut guard) => {
+                                        tracing::debug!(fd, "I/O ready to write");
+                                        unsafe {
+                                            // libevent dispatches the callback with the mapped file descriptor
+                                            libevent_sys::evmap_io_active_(base.0, fd, libevent_sys::EV_WRITE as i16);
+                                        }
+
+                                        // If the ready flag is not cleared, then this loop will hang the runtime.
+                                        guard.clear_ready();
+                                    }
+                                    Err(error) => {
+                                        tracing::error!(?error);
+                                        break;
+                                    }
+                                }
+                            },
+                            _ = notify.notified() => break, // terminate the task on notification
+                        }
+                    }
+
+                    tracing::debug!(?fd, "I/O task terminated");
+                });
+
+                0
+            }
+            Err(error) => {
+                tracing::error!(?error);
+                -1
+            }
+        }
+    }
+
+    /// Terminates an active I/O task
+    #[instrument]
+    fn del_io(&self, fdinfo: *mut Arc<Notify>) -> c_int {
+        tracing::debug!("delete an I/O event");
+
+        unsafe {
+            match fdinfo.as_mut() {
+                Some(notify) => {
+                    notify.notify_one();
+                    std::ptr::drop_in_place(fdinfo);
+                    0
+                }
+                None => -1,
+            }
+        }
+    }
+
+    /// Drive the tokio runtime with an optional duration for timout events
+    #[instrument]
+    fn dispatch(&self, timeout: Option<Duration>) {
+        self.runtime.block_on(async move {
+            match timeout {
+                // spawned tasks are serviced during the sleep time
+                Some(timeout) => tokio::time::sleep(timeout).await,
```

Comment:
The issue I see with this approach is that an I/O event that libevent has registered with tokio might result in timers being registered that would then be at the top of the timer heap.  This implementation would not return control back to libevent to possibly execute these actions by the deadlines for those new timers.

Take for example the following scenario:

1. I register a 30s timer with libevent
2. Libevent calls dispatch, there are 29.99s until the next timer needs to fire so that is the timeout passed into dispatch
3. After 0.99s an I/O event occurs and a singleshot 250ms timer is registered with libevent.
4. The remaining 29s elapses and libevent sees that it needs to service the 250ms oneshot timer but we've done so 28.75s late of its intended deadline.

I think this situation could occur with this implementation.

Reply:
I think that you are right. We could use a Notify to break out of the loop, but that seems like a big performance hit. I need to do some profiling.

Reply:
I modified the `time-test` sample program to verify this. I extended the timeout to 10 seconds. Then I set an SIGALRM handler for 1 second (so it doesn't use libevent timers). In that handler, I set another 0 second timeout that then sets up another SIGALRM.

The original version of this PR delayed the signal handling and 0 second timeout until the initial 10 second timeout completed. That's not executable. So, I added a notification mechanism to allow the block_on to break out early if an event occurs.

With the change, the SIGALRM is handled every second and the 0 second timeout occurs immediately afterwards.

Reply:
![libevent-benchmark](https://user-images.githubusercontent.com/468066/135874112-25f90e18-51a6-4fd6-86ce-1285a9987299.png)

I adapted `libevent/test/bench.c` to run with the tokio backend. It is definitely slower than using the native kqueue backend on my MacBook.

Reply:
![kqueue-vs-tokio](https://user-images.githubusercontent.com/468066/136150068-c1b92a58-50f0-479a-a92f-3aa2d7e812bb.png)

I haven't been able to get the performance on this benchmark any closer than half as good as straight kqueue. We could still end up doing better in practice because of better coordination between tokio and libevent.